In [1]:
import pandas as pd

In [2]:
cforum = pd.read_csv('Cforum_2016.csv')
aforum = pd.read_csv('Aforum_2016.csv')


Merging Aforum and Cforum into a single dataframe: df and adding a column 'Forum' to show forum name.

In [3]:
df = pd.concat([cforum, aforum])
df

Link  \
0       https://www.canadavisa.com//canada-immigration...   
1       https://www.canadavisa.com//canada-immigration...   
2       https://www.canadavisa.com//canada-immigration...   
3       https://www.canadavisa.com//canada-immigration...   
4       https://www.canadavisa.com//canada-immigration...   
...                                                   ...   
106452  https://www.australiaforum.com/visas-immigrati...   
106453  https://www.australiaforum.com/visas-immigrati...   
106454  https://www.australiaforum.com/visas-immigrati...   
106455  https://www.australiaforum.com/visas-immigrati...   
106456  https://www.australiaforum.com/visas-immigrati...   

                       Date         Username  \
0       2019-09-30 00:00:00  Prashant-Canada   
1       2019-06-11 00:00:00      spanishnick   
2       2020-01-22 00:00:00     Junaid yasin   
3       2020-01-22 00:00:00     Junaid yasin   
4       2020-01-23 00:00:00     canuck_in_uk   
...                     ...              ...   
106452  2016-01-02 03:24:00     xenaclaire86   
106453  2016-01-01 02:51:00      sunflowerhk   
106454  2016-01-01 04:39:00     Maggie-May24   
106455  2016-01-01 03:14:00        Tahir1980   
106456  2016-01-01 00:40:00     Maggie-May24   

                                                  Message    Country  Like  \
0       I’m filing  Switzerland Visa for my family as ...        NaN     0   
1       Hi,<br/><br/>My aunt and uncle applied for a t...        NaN     0   
2                                     I need job in canda        NaN     0   
3       <blockquote class="bbCodeBlock bbCodeBlock--ex...        NaN     0   
4       can possible without IELTS test LMIA i am from...        NaN     1   
...                                                   ...        ...   ...   
106452  \r\n\t\t\t\r\n\t\t\tThanks for the replies mat...        NaN     0   
106453  \r\n\t\t\t\r\n\t\t\tA few questions about how ...  hong_kong     0   
106454  \r\n\t\t\t\r\n\t\t\tYou can find information a...     canada  1522   
106455  \r\n\t\t\t\r\n\t\t\tI have applied for 189 vis...   pakistan     0   
106456  \r\n\t\t\t\r\n\t\t\tDIBP publishes this on the...     canada  1522   

                              Topic                                  Thread  \
0       Visa Offices in Europe\nNew                           Schengen VISA   
1       Visa Offices in Europe\nNew                         Visa from Paris   
2       Visa Offices in Europe\nNew              Need job and visa in canda   
3       Visa Offices in Europe\nNew              Need job and visa in canda   
4       Visa Offices in Europe\nNew              Need job and visa in canda   
...                             ...                                     ...   
106452        Visas and immigration         Student visa to spouse visa 820   
106453        Visas and immigration            RRV/citizenship requirements   
106454        Visas and immigration            RRV/citizenship requirements   
106455        Visas and immigration                      Medical of newborn   
106456        Visas and immigration  Submitted EOI, waiting for invitation.   

                                        Message_lowercase  \
0       i’m filing switzerland visa for my family as a...   
1       hi,<br/><br/>my aunt and uncle applied for a t...   
2                                     i need job in canda   
3       <blockquote class="bbcodeblock bbcodeblock--ex...   
4       can possible without ielts test lmia i am from...   
...                                                   ...   
106452  thanks for the replies mates. cancellation mea...   
106453  a few questions about how to get rrv or citize...   
106454  you can find information about both rrvs and c...   
106455  i have applied for 189 visa. i am primary appl...   
106456  dibp publishes this on their website where you...   

                                        Message_no_quotes month_year  \
0       i’m filing switzerland visa for my family as

In [4]:
aforum['forum'] = pd.Series(['Aforum' for x in range(len(df.index))])
cforum['forum'] = pd.Series(['Cforum' for x in range(len(df.index))])


In [5]:
df = pd.concat([cforum, aforum])

Adding a year column.

In [6]:
df['Date'] = pd.to_datetime(df['Date'], errors="coerce")
df['Date'].values.astype('datetime64[Y]')
df['year'] =  pd.DatetimeIndex(df['Date']).year

In [7]:
df['year']

0         2019.0
1         2019.0
2         2020.0
3         2020.0
4         2020.0
           ...  
106452    2016.0
106453    2016.0
106454    2016.0
106455    2016.0
106456    2016.0
Name: year, Length: 200669, dtype: float64

Cleaning messages. Some inconistencies with the 2 datasets. So need to do a fresh cleanup. 

In [8]:
df = df.drop(columns=['Message_lowercase','Message_nopunct', 'Message_no_quotes'])
df.columns

Index(['Link', 'Date', 'Username', 'Message', 'Country', 'Like', 'Topic',
       'Thread', 'month_year', 'forum', 'year'],
      dtype='object')

 "blockquote" represents duplicate messages. Need to remove from all messages.

In [9]:
df = df[df['Message'].notnull()]


In [10]:
#coverting first to lowercase
df['Message_lowercase'] = df['Message'].apply(lambda x: " ".join(x.lower() for x in x.split()))
df.head()


Link       Date  \
0  https://www.canadavisa.com//canada-immigration... 2019-09-30   
1  https://www.canadavisa.com//canada-immigration... 2019-06-11   
2  https://www.canadavisa.com//canada-immigration... 2020-01-22   
3  https://www.canadavisa.com//canada-immigration... 2020-01-22   
4  https://www.canadavisa.com//canada-immigration... 2020-01-23   

          Username                                            Message Country  \
0  Prashant-Canada  I’m filing  Switzerland Visa for my family as ...     NaN   
1      spanishnick  Hi,<br/><br/>My aunt and uncle applied for a t...     NaN   
2     Junaid yasin                                I need job in canda     NaN   
3     Junaid yasin  <blockquote class="bbCodeBlock bbCodeBlock--ex...     NaN   
4     canuck_in_uk  can possible without IELTS test LMIA i am from...     NaN   

  Like                        Topic                      Thread month_year  \
0    0  Visa Offices in Europe\nNew               Schengen VISA    2019-09   
1    0  Visa Offices in Europe\nNew             Visa from Paris    2019-06   
2    0  Visa Offices in Europe\nNew  Need job and visa in canda    2020-01   
3    0  Visa Offices in Europe\nNew  Need job and visa in canda    2020-01   
4    1  Visa Offices in Europe\nNew  Need job and visa in canda    2020-01   

    forum    year                                  Message_lowercase  
0  Cforum  2019.0  i’m filing switzerland visa for my family as a...  
1  Cforum  2019.0  hi,<br/><br/>my aunt and uncle applied for a t...  
2  Cforum  2020.0                                i need job in canda  
3  Cforum  2020.0  <blockquote class="bbcodeblock bbcodeblock--ex...  
4  Cforum  2020.0  can possible without ielts test lmia i am from...

In [11]:
#creating a function to remove messages stored in blockquote
from bs4 import BeautifulSoup as bs

def deblock(message):
    mbs = bs(message, 'html.parser')
    for bq in mbs.findAll('blockquote'):
        bq.decompose()
    return str(mbs)

LM's insertions

In [28]:
#creating a function to remove messages stored in blockquote
from bs4 import BeautifulSoup as bs

def deblock_lm(message):
    mbs = bs(message, 'html.parser')
    for bq in mbs.findAll('blockquote'):
        bq.decompose()
    # Note this change
    return mbs.text

In [29]:
df['Message_no_quotes_lm'] = df['Message_lowercase'].apply(lambda x: deblock_lm(x))

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://www.change.org/p/minister-of-immigration-diversity-and-inclusiveness-quebec-online-fraud-by-midi-new-rule-must-not-be-applicable-to-applicants-before-8th-march-2017?recruiter=693592760&amp;utm_source=share_petition&amp;utm_medium=copylink" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to get the document behind the URL, and feed that document to Beautiful Soup.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://chat.whatsapp.com/j9gfjmrxenxljmniqaiykx" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to get the document behind the URL, and feed that document to Beautiful Soup.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://www.thestar.com/news/immigration/2016/08/18/ottawas-new-air-travel-rule-catches-dual-citizens-by-surprise.html" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to get the document behind the URL, and feed that document to Beautiful Soup.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "http://www.cic.gc.ca/english/work/" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to get the document behind the URL, and feed that document to Beautiful Soup.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "http://www.ontarioimmigration.ca/en/working/oi_how_work_engineer.html" looks like a URL. Beautifu

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://immi.homeaffairs.gov.au/visa...n-graduate-476" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to get the document behind the URL, and feed that document to Beautiful Soup.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://www.modernghana.com/amp/news...rom%20%251%24s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to get the document behind the URL, and feed that document to Beautiful Soup.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://www.vetassess.com.au/portals...tylecturer.pdf" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP 

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://www.homeaffairs.gov.au/trav/...ans-fact-sheet" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to get the document behind the URL, and feed that document to Beautiful Soup.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://www.ombudsman.gov.au/" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to get the document behind the URL, and feed that document to Beautiful Soup.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://www.change.org/p/peter-dutto..._starters_page" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to 

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://www.homeaffairs.gov.au/trav/...#tab-content-4" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to get the document behind the URL, and feed that document to Beautiful Soup.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://www.homeaffairs.gov.au/visas...mployment.aspx" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to get the document behind the URL, and feed that document to Beautiful Soup.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "http://www.australiaforum.com/visas-...isa-309-a.html" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP 

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://www.border.gov.au/about/cont...ts-suggestions" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to get the document behind the URL, and feed that document to Beautiful Soup.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://www.dropbox.com/s/n7opjs279u...52.51.png?dl=0" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like requests to get the document behind the URL, and feed that document to Beautiful Soup.
  warnings.warn(
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/bs4/__init__.py:393: UserWarning: "https://www.border.gov.au/trav/visa-1/155-" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client like

In [14]:
df['Message_no_quotes'] = df['Message_lowercase'].apply(lambda x: deblock(x))


In [30]:
df['Message_no_quotes_lm'].head()

0    i’m filing switzerland visa for my family as a tourist .i have an appointment for finger printin...
1    hi,my aunt and uncle applied for a tourist visa at the canadian embassy in paris more than five ...
2                                                                                    i need job in canda
3                https://www.canadavisa.com/canada-immigration-discussion-board/threads/need-job.674699/
4    can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...
Name: Message_no_quotes_lm, dtype: object

In [31]:
#removing punct
import string
#widen the column
pd.set_option('display.max_colwidth', 100)


df['Message_punct'] = df['Message_lowercase'].apply(lambda x:''.join([i for i in x if i not in string.punctuation]))

# Why not use the no quotes version?
df['Message_punct_lm'] = df['Message_no_quotes_lm'].apply(lambda x:''.join([i for i in x if i not in string.punctuation]))

df.head()

Link  \
0  https://www.canadavisa.com//canada-immigration-discussion-board/threads/schengen-visa.656723/pos...   
1  https://www.canadavisa.com//canada-immigration-discussion-board/threads/visa-from-paris.636790/p...   
2  https://www.canadavisa.com//canada-immigration-discussion-board/threads/need-job-and-visa-in-can...   
3  https://www.canadavisa.com//canada-immigration-discussion-board/threads/need-job-and-visa-in-can...   
4  https://www.canadavisa.com//canada-immigration-discussion-board/threads/need-job-and-visa-in-can...   

        Date         Username  \
0 2019-09-30  Prashant-Canada   
1 2019-06-11      spanishnick   
2 2020-01-22     Junaid yasin   
3 2020-01-22     Junaid yasin   
4 2020-01-23     canuck_in_uk   

                                                                                               Message  \
0  I’m filing  Switzerland Visa for my family as a tourist .<br/><br/><br/><ul><li data-xf-list-typ...   
1  Hi,<br/><br/>My aunt and uncle applied for a tourist visa at the Canadian embassy in Paris more ...   
2                                                                                  I need job in canda   
3  <blockquote class="bbCodeBlock bbCodeBlock--expandable bbCodeBlock--quote"><div class="bbCodeBlo...   
4  can possible without IELTS test LMIA i am from Pakistan but working Saudi Arabia on Riyadh Metro...   

  Country Like                        Topic                      Thread  \
0     NaN    0  Visa Offices in Europe\nNew               Schengen VISA   
1     NaN    0  Visa Offices in Europe\nNew             Visa from Paris   
2     NaN    0  Visa Offices in Europe\nNew  Need job and visa in canda   
3     NaN    0  Visa Offices in Europe\nNew  Need job and visa in canda   
4     NaN    1  Visa Offices in Europe\nNew  Need job and visa in canda   

  month_year   forum    year  \
0    2019-09  Cforum  2019.0   
1    2019-06  Cforum  2019.0   
2    2020-01  Cforum  2020.0   
3    2020-01  Cforum  2020.0   
4    2020-01  Cforum  2020.0   

                                                                                     Message_lowercase  \
0  i’m filing switzerland visa for my family as a tourist .<br/><br/><br/><ul><li data-xf-list-type...   
1  hi,<br/><br/>my aunt and uncle applied for a tourist visa at the canadian embassy in paris more ...   
2                                                                                  i need job in canda   
3  <blockquote class="bbcodeblock bbcodeblock--expandable bbcodeblock--quote"><div class="bbcodeblo...   
4  can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...   

                                                                                  Message_no_quotes_lm  \
0  i’m filing switzerland visa for my family as a tourist .i have an appointment for finger printin...   
1  hi,my aunt and uncle applied for a tourist visa at the canadian embassy in paris more than five ...   
2                                                                                  i need job in canda   
3              https://www.canadavisa.com/canada-immigration-discussion-board/threads/need-job.674699/   
4  can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...   

                                                                                     Message_no_quotes  \
0  i’m filing switzerland visa for my family as a tourist .<br/><br/><br/><ul><li data-xf-list-type...   
1  hi,<br/><br/>my aunt and uncle applied for a tourist visa at the canadian embassy in paris more ...   
2                                                                                  i need job in canda   
3  <a class="link link--internal" href="https://www.canadavisa.com/canada-immigration-discussion-bo...   
4  can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...   

                                                                      

Using string.printable to filter out special chars. 

In [60]:


#df.Message_punct.str.replace('[^a-zA-Z]', '')

from string import printable
st = set(printable)
#df['Message_punct'] = df["Message_punct"].apply(lambda x: ''.join([" " if  i not in  st else i for i in x]))
df['Message_punct_lm'] = df["Message_punct_lm"].apply(lambda x: ''.join([" " if  i not in  st else i for i in x]))

In [17]:
df['Message_punct'].head()

0    i’m filing switzerland visa for my family as a tourist brbrbrulli dataxflisttypeuli have an appo...
1    hibrbrmy aunt and uncle applied for a tourist visa at the canadian embassy in paris more than fi...
2                                                                                    i need job in canda
3    blockquote classbbcodeblock bbcodeblockexpandable bbcodeblockquotediv classbbcodeblocktitlea cla...
4    can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...
Name: Message_punct, dtype: object

In [18]:
df['Message_punct_lm'].head()

0    i’m filing switzerland visa for my family as a tourist i have an appointment for finger printing...
1    himy aunt and uncle applied for a tourist visa at the canadian embassy in paris more than five w...
2                                                                                    i need job in canda
3                              httpswwwcanadavisacomcanadaimmigrationdiscussionboardthreadsneedjob674699
4    can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...
Name: Message_punct_lm, dtype: object

In [19]:
#The dtm has several non-english words
import nltk
import ssl

try:
    _create_unverified_https_context = ssl._create_unverified_context
except AttributeError:
    pass
else:
    ssl._create_default_https_context = _create_unverified_https_context
    
nltk.download('words')
words = df['Message_punct']
words.head()

[nltk_data] Downloading package words to /Users/liam/nltk_data...
[nltk_data]   Package words is already up-to-date!


0    i’m filing switzerland visa for my family as a tourist brbrbrulli dataxflisttypeuli have an appo...
1    hibrbrmy aunt and uncle applied for a tourist visa at the canadian embassy in paris more than fi...
2                                                                                    i need job in canda
3    blockquote classbbcodeblock bbcodeblockexpandable bbcodeblockquotediv classbbcodeblocktitlea cla...
4    can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...
Name: Message_punct, dtype: object

In [20]:
df.head()

Link  \
0  https://www.canadavisa.com//canada-immigration-discussion-board/threads/schengen-visa.656723/pos...   
1  https://www.canadavisa.com//canada-immigration-discussion-board/threads/visa-from-paris.636790/p...   
2  https://www.canadavisa.com//canada-immigration-discussion-board/threads/need-job-and-visa-in-can...   
3  https://www.canadavisa.com//canada-immigration-discussion-board/threads/need-job-and-visa-in-can...   
4  https://www.canadavisa.com//canada-immigration-discussion-board/threads/need-job-and-visa-in-can...   

        Date         Username  \
0 2019-09-30  Prashant-Canada   
1 2019-06-11      spanishnick   
2 2020-01-22     Junaid yasin   
3 2020-01-22     Junaid yasin   
4 2020-01-23     canuck_in_uk   

                                                                                               Message  \
0  I’m filing  Switzerland Visa for my family as a tourist .<br/><br/><br/><ul><li data-xf-list-typ...   
1  Hi,<br/><br/>My aunt and uncle applied for a tourist visa at the Canadian embassy in Paris more ...   
2                                                                                  I need job in canda   
3  <blockquote class="bbCodeBlock bbCodeBlock--expandable bbCodeBlock--quote"><div class="bbCodeBlo...   
4  can possible without IELTS test LMIA i am from Pakistan but working Saudi Arabia on Riyadh Metro...   

  Country Like                        Topic                      Thread  \
0     NaN    0  Visa Offices in Europe\nNew               Schengen VISA   
1     NaN    0  Visa Offices in Europe\nNew             Visa from Paris   
2     NaN    0  Visa Offices in Europe\nNew  Need job and visa in canda   
3     NaN    0  Visa Offices in Europe\nNew  Need job and visa in canda   
4     NaN    1  Visa Offices in Europe\nNew  Need job and visa in canda   

  month_year   forum    year  \
0    2019-09  Cforum  2019.0   
1    2019-06  Cforum  2019.0   
2    2020-01  Cforum  2020.0   
3    2020-01  Cforum  2020.0   
4    2020-01  Cforum  2020.0   

                                                                                     Message_lowercase  \
0  i’m filing switzerland visa for my family as a tourist .<br/><br/><br/><ul><li data-xf-list-type...   
1  hi,<br/><br/>my aunt and uncle applied for a tourist visa at the canadian embassy in paris more ...   
2                                                                                  i need job in canda   
3  <blockquote class="bbcodeblock bbcodeblock--expandable bbcodeblock--quote"><div class="bbcodeblo...   
4  can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...   

                                                                                  Message_no_quotes_lm  \
0  i’m filing switzerland visa for my family as a tourist .i have an appointment for finger printin...   
1  hi,my aunt and uncle applied for a tourist visa at the canadian embassy in paris more than five ...   
2                                                                                  i need job in canda   
3              https://www.canadavisa.com/canada-immigration-discussion-board/threads/need-job.674699/   
4  can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...   

                                                                                     Message_no_quotes  \
0  i’m filing switzerland visa for my family as a tourist .<br/><br/><br/><ul><li data-xf-list-type...   
1  hi,<br/><br/>my aunt and uncle applied for a tourist visa at the canadian embassy in paris more ...   
2                                                                                  i need job in canda   
3  <a class="link link--internal" href="https://www.canadavisa.com/canada-immigration-discussion-bo...   
4  can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...   

                                                                      

In [32]:
import pickle
df.to_pickle("corpus.pkl")

Message_punct is the cleaned column to work with (still might need some cleaning)

In [33]:
df['Message_lowercase'].head()

0    i’m filing switzerland visa for my family as a tourist .<br/><br/><br/><ul><li data-xf-list-type...
1    hi,<br/><br/>my aunt and uncle applied for a tourist visa at the canadian embassy in paris more ...
2                                                                                    i need job in canda
3    <blockquote class="bbcodeblock bbcodeblock--expandable bbcodeblock--quote"><div class="bbcodeblo...
4    can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...
Name: Message_lowercase, dtype: object

In [21]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(stop_words='english')
data_cv = cv.fit_transform(df['Message_punct'])
data_dtm = pd.DataFrame(data_cv.toarray(), columns=cv.get_feature_names())
data_dtm.index = df.index
#data_dtm

In [33]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(stop_words='english')
data_cv = cv.fit_transform(df['Message_punct_lm'])
data_dtm = pd.DataFrame(data_cv.toarray(), columns=cv.get_feature_names())
data_dtm.index = df.index
#data_dtm

This is the complete dtm for use. first few iters had lots of non-english chars so tried re.sub but didnt work. The string.printable method seems to work. Pickling results in dead kernel every time. 

## Exploratory Data Analysis

In [25]:
data = pd.read_pickle('corpus.pkl')

Creating and pickling slices of data by forum and year to compare over time. 

In [49]:
dfA = data.loc[df['forum'] == 'Aforum']
dfA.to_pickle("Aforum.pkl")
dfC = data.loc[df['forum'] == 'Cforum']
dfC.to_pickle("Cforum.pkl")
y2016A = dfA[dfA.year == 2016]
y2016A.to_pickle("y2016A.pkl")
y2017A = dfA[dfA.year == 2017]
y2017A.to_pickle("y2017A.pkl")
y2018A = dfA[dfA.year == 2018]
y2018A.to_pickle("y2018A.pkl")
y2019A = dfA[dfA.year == 2019]
y2019A.to_pickle("y2019A.pkl")
y2020A = dfA[dfA.year == 2020]
y2020A.to_pickle("y2020A.pkl")



In [26]:
y2016C = dfC[dfC.year == 2016]
y2016C.to_pickle("y2016C.pkl")
y2017C = dfC[dfC.year == 2017]
y2017C.to_pickle("y2017C.pkl")
y2018C = dfC[dfC.year == 2018]
y2018C.to_pickle("y2018C.pkl")
y2019C = dfC[dfC.year == 2019]
y2019C.to_pickle("y2019C.pkl")
y2020C = dfC[dfC.year == 2020]
y2020C.to_pickle("y2020C.pkl")
y2020C.head()

Link  \
2     https://www.canadavisa.com//canada-immigration-discussion-board/threads/need-job-and-visa-in-can...   
3     https://www.canadavisa.com//canada-immigration-discussion-board/threads/need-job-and-visa-in-can...   
4     https://www.canadavisa.com//canada-immigration-discussion-board/threads/need-job-and-visa-in-can...   
1179  https://www.canadavisa.com//canada-immigration-discussion-board/threads/how-to-check-lmia-letter...   
1180  https://www.canadavisa.com//canada-immigration-discussion-board/threads/how-to-check-lmia-letter...   

           Date        Username  \
2    2020-01-22    Junaid yasin   
3    2020-01-22    Junaid yasin   
4    2020-01-23    canuck_in_uk   
1179 2020-01-03    canuck_in_uk   
1180 2020-01-03  Sandip Shyamal   

                                                                                                  Message  \
2                                                                                     I need job in canda   
3     <blockquote class="bbCodeBlock bbCodeBlock--expandable bbCodeBlock--quote"><div class="bbCodeBlo...   
4     can possible without IELTS test LMIA i am from Pakistan but working Saudi Arabia on Riyadh Metro...   
1179  OFFER LETTER<br/><br/>Date of issue: 07/01/2020<br/><br/>I am pleased to confirm our offer of em...   
1180  <blockquote class="bbCodeBlock bbCodeBlock--expandable bbCodeBlock--quote"><div class="bbCodeBlo...   

     Country Like                             Topic  \
2        NaN    0       Visa Offices in Europe\nNew   
3        NaN    0       Visa Offices in Europe\nNew   
4        NaN    1       Visa Offices in Europe\nNew   
1179     NaN    0  Canadian Labour Legislation\nNew   
1180     NaN    0  Canadian Labour Legislation\nNew   

                          Thread month_year   forum    year  \
2     Need job and visa in canda    2020-01  Cforum  2020.0   
3     Need job and visa in canda    2020-01  Cforum  2020.0   
4     Need job and visa in canda    2020-01  Cforum  2020.0   
1179    how to check LMIA letter    2020-01  Cforum  2020.0   
1180    how to check LMIA letter    2020-01  Cforum  2020.0   

                                                                                        Message_lowercase  \
2                                                                                     i need job in canda   
3     <blockquote class="bbcodeblock bbcodeblock--expandable bbcodeblock--quote"><div class="bbcodeblo...   
4     can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...   
1179  offer letter<br/><br/>date of issue: 07/01/2020<br/><br/>i am pleased to confirm our offer of em...   
1180  <blockquote class="bbcodeblock bbcodeblock--expandable bbcodeblock--quote"><div class="bbcodeblo...   

                                                                                        Message_no_quotes  \
2                                                                                     i need job in canda   
3     <a class="link link--internal" href="https://www.canadavisa.com/canada-immigration-discussion-bo...   
4     can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...   
1179  offer letter<br/><br/>date of issue: 07/01/2020<br/><br/>i am pleased to confirm our offer of em...   
1180  it is fake there will be no job. same letter posted by someone else a few days ago and pointed o...   

                                                                                            Message_punct  
2                                                                                     i need job in canda  
3     blockquote classbbcodeblock bbcodeblockexpandable bbcodeblockquotediv classbbcodeblocktitlea cla...  
4     can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...  
1179  offer letterbrbrdate of issue 07012020brbri am pleased to confirm our offer of employment to you...  
1180  blockquote cla

In [26]:
commonwords = pd.Series(' '.join(data.Message_punct).split()).value_counts()[:15]
commonwords

the     553237
to      518523
i       386676
and     327778
a       302047
for     279344
you     261525
in      259456
is      219867
of      211217
my      189737
it      162829
have    161064
that    156199
on      135524
dtype: int64

In [34]:
commonwords = pd.Series(' '.join(data.Message_punct_lm).split()).value_counts()[:15]
commonwords

the     473672
to      395553
i       332729
and     278181
a       258833
for     238927
you     221480
in      216345
is      185917
of      178584
my      159019
it      140188
have    138425
that    133960
visa    122960
dtype: int64

In [35]:
from sklearn.feature_extraction import text
from sklearn.feature_extraction.text import CountVectorizer
#add new stopwords
stop_words = text.ENGLISH_STOP_WORDS.union(commonwords)

# Recreate document-term matrix which excludes additional stop words
cv = CountVectorizer(stop_words=stop_words)
data_cv = cv.fit_transform(data.Message_punct)
dtm_stop = pd.DataFrame(data_cv.toarray(), columns=cv.get_feature_names())
dtm_stop.index = data.index

dtm_stop

00  000  0000  00000  000000  0000000  00000000  000000000  \
0        0    0     0      0       0        0         0          0   
1        0    0     0      0       0        0         0          0   
2        0    0     0      0       0        0         0          0   
3        0    0     0      0       0        0         0          0   
4        0    0     0      0       0        0         0          0   
...     ..  ...   ...    ...     ...      ...       ...        ...   
106452   0    0     0      0       0        0         0          0   
106453   0    0     0      0       0        0         0          0   
106454   0    0     0      0       0        0         0          0   
106455   0    0     0      0       0        0         0          0   
106456   0    0     0      0       0        0         0          0   

        0000000000  00000000000000  ...  यव  वर  शन  २०१६  २८  സർ  ᧗ࡦᆳⲵᮠ䟿ഐѫ  \
0                0               0  ...   0   0   0     0   0   0         0   
1                0               0  ...   0   0   0     0   0   0         0   
2                0               0  ...   0   0   0     0   0   0         0   
3                0               0  ...   0   0   0     0   0   0         0   
4                0               0  ...   0   0   0     0   0   0         0   
...            ...             ...  ...  ..  ..  ..   ...  ..  ..       ...   
106452           0               0  ...   0   0   0     0   0   0         0   
106453           0               0  ...   0   0   0     0   0   0         0   
106454           0               0  ...   0   0   0     0   0   0         0   
106455           0               0  ...   0   0   0     0   0   0         0   
106456           0               0  ...   0   0   0     0   0   0         0   

        ⲵ䙊䗷ӱѹᇓր  䇽ˈ  䘉њ  
0             0   0   0  
1             0   0   0  
2             0   0   0  
3             0   0   0  
4             0   0   0  
...         ...  ..  ..  
106452        0   0   0  
106453        0   0   0  
106454        0   0   0  
106455        0   0   0  
106456        0   0   0  

[200669 rows x 283843 columns]

In [36]:
from sklearn.feature_extraction import text
from sklearn.feature_extraction.text import CountVectorizer
#add new stopwords
stop_words = text.ENGLISH_STOP_WORDS.union(commonwords)

# Recreate document-term matrix which excludes additional stop words
cv = CountVectorizer(stop_words=stop_words)
data_cv = cv.fit_transform(data.Message_punct_lm)
dtm_stop = pd.DataFrame(data_cv.toarray(), columns=cv.get_feature_names())
dtm_stop.index = data.index

#dtm_stop

In [42]:
import numpy as np

In [44]:
sum_words = data_cv.sum(axis=0)

In [92]:
data_cv_au = cv.fit_transform(data.loc[df['forum'] == 'Aforum'].Message_punct_lm)



In [102]:
data_cv_cn = cv.fit_transform(data.loc[df['forum'] == 'Cforum'].Message_punct_lm)

In [93]:
sum_words_au = data_cv_au.sum(axis=0)
words_freq_au = [(word, sum_words_au[0, idx]) for word, idx in cv.vocabulary_.items()]
words_freq_au =sorted(words_freq_au, key = lambda x: x[1], reverse=True)

In [103]:
sum_words_cn = data_cv_cn.sum(axis=0)
words_freq_cn = [(word, sum_words_cn[0, idx]) for word, idx in cv.vocabulary_.items()]
words_freq_cn =sorted(words_freq_cn, key = lambda x: x[1], reverse=True)

In [104]:
sum_words.sum()

6254843

In [105]:
sum_words_au.sum()

4386372

In [106]:
sum_words_cn.sum()

1868471

In [98]:
3960850

3960850

In [107]:
words_freq

[('visa', 122988),
 ('application', 67614),
 ('just', 45918),
 ('months', 44891),
 ('posted', 43702),
 ('quote', 43525),
 ('originally', 42184),
 ('time', 40941),
 ('need', 35472),
 ('australia', 34496),
 ('apply', 34338),
 ('partner', 31613),
 ('hi', 31572),
 ('know', 29111),
 ('thanks', 28913),
 ('applied', 27635),
 ('got', 27334),
 ('did', 26310),
 ('years', 25166),
 ('dont', 24901),
 ('work', 24727),
 ('im', 23288),
 ('canada', 20952),
 ('like', 20771),
 ('year', 19531),
 ('pr', 19504),
 ('granted', 19092),
 ('waiting', 18728),
 ('case', 18661),
 ('date', 17825),
 ('help', 17810),
 ('received', 17695),
 ('long', 17677),
 ('think', 17387),
 ('good', 17326),
 ('documents', 17201),
 ('does', 16985),
 ('relationship', 16695),
 ('thank', 16611),
 ('new', 16537),
 ('days', 15951),
 ('evidence', 15851),
 ('information', 15834),
 ('want', 15755),
 ('immigration', 15312),
 ('process', 15272),
 ('form', 15263),
 ('email', 15260),
 ('people', 15122),
 ('processing', 14944),
 ('country', 14934

In [108]:
words_freq_au

[('visa', 115068),
 ('application', 48968),
 ('quote', 43429),
 ('posted', 43199),
 ('originally', 42074),
 ('months', 38043),
 ('just', 35059),
 ('australia', 34210),
 ('partner', 30916),
 ('time', 29802),
 ('apply', 25490),
 ('need', 24480),
 ('hi', 24020),
 ('thanks', 22243),
 ('applied', 21823),
 ('know', 20346),
 ('years', 18928),
 ('granted', 18735),
 ('im', 18457),
 ('did', 17985),
 ('dont', 17378),
 ('got', 17075),
 ('relationship', 16105),
 ('work', 15841),
 ('evidence', 15331),
 ('like', 14688),
 ('year', 14647),
 ('waiting', 14309),
 ('820', 14132),
 ('case', 14019),
 ('long', 13365),
 ('date', 13126),
 ('documents', 12738),
 ('does', 12513),
 ('think', 12483),
 ('help', 12224),
 ('thank', 12196),
 ('good', 12189),
 ('agent', 12098),
 ('form', 12031),
 ('new', 11889),
 ('australian', 11750),
 ('immigration', 11531),
 ('information', 11431),
 ('want', 10885),
 ('grant', 10854),
 ('processing', 10839),
 ('people', 10815),
 ('wait', 10741),
 ('801', 10695),
 ('police', 10581),


In [109]:
words_freq_cn

[('canada', 19718),
 ('application', 18646),
 ('time', 11139),
 ('pr', 11079),
 ('need', 10992),
 ('just', 10859),
 ('got', 10259),
 ('work', 8886),
 ('apply', 8848),
 ('know', 8765),
 ('received', 8714),
 ('did', 8325),
 ('visa', 7920),
 ('hi', 7552),
 ('dont', 7523),
 ('months', 6848),
 ('thanks', 6670),
 ('canadian', 6250),
 ('process', 6247),
 ('years', 6238),
 ('like', 6083),
 ('citizenship', 6007),
 ('test', 5995),
 ('status', 5930),
 ('applied', 5812),
 ('days', 5757),
 ('help', 5586),
 ('letter', 5463),
 ('job', 5450),
 ('cic', 5334),
 ('passport', 5192),
 ('country', 5154),
 ('good', 5137),
 ('sent', 5105),
 ('permit', 4957),
 ('think', 4904),
 ('year', 4884),
 ('want', 4870),
 ('im', 4831),
 ('medical', 4808),
 ('email', 4782),
 ('date', 4699),
 ('new', 4648),
 ('case', 4642),
 ('yes', 4565),
 ('send', 4531),
 ('does', 4472),
 ('documents', 4463),
 ('waiting', 4419),
 ('thank', 4415),
 ('information', 4403),
 ('long', 4312),
 ('people', 4307),
 ('question', 4299),
 ('sure', 4

I am kind of stuck with the dtm at this point because I could do further analysis on terms but it only seems to work based on columns (i.e term-document matrix). So for example, for each conversation, on the forum, I could get the top n words. I guess that could be converted into a dataframe for further analysis?

A very quick word freq of each forum but not sure why the (commonwords) stopwords are still there. Also Cforum still needs cleaning.

In [51]:
freqA = pd.Series(' '.join(dfA.Message_punct).split()).value_counts()[:200]
freqC = pd.Series(' '.join(dfC.Message_punct).split()).value_counts()[:200]
print(freqC)
pd.set_option('display.max_rows', None)

to                                                                            241640
the                                                                           208861
i                                                                             151190
and                                                                           128547
in                                                                            115248
you                                                                           114990
a                                                                             107793
for                                                                           107644
is                                                                             90726
of                                                                             85058
my                                                                             78072
it                                                               

In [31]:
freqC = pd.Series(' '.join(dfC.Message_punct).split()).value_counts()[:200]
print(freqA)
pd.set_option('display.max_rows', None)

the             344473
to              276903
i               239225
and             199246
a               194305
for             171719
you             147149
in              144267
is              129149
of              126161
visa            115132
my              111680
it              103702
have            100267
that             98260
on               81745
be               76224
we               73222
as               64362
if               64058
not              61412
by               60956
your             60025
this             57970
are              55051
with             51406
but              50571
they             50532
application      49016
will             48710
or               48606
from             48282
can              46584
was              45645
so               45157
quote            43429
posted           43199
originally       42074
months           38056
do               37565
all              35387
an               35369
just             35060
would      

## Initial Findings
Just exploring the top 200 words in both forums - common words include 'application', 'work', 'wait'. 
But there are some words that are somewhat unique to each forum - such as 'relationship', 'partner' and '309' in Australia -- all referring to partner visas which is what a large number of conversations are about. 
For Cforum, words like 'permit', 'card' and 'letter' appear more frequently, referring to employment and work permits, and work experience letters. 



## Sentiment Analysis

In [44]:
from textblob import TextBlob

pol = lambda x: TextBlob(x).sentiment.polarity
sub = lambda x: TextBlob(x).sentiment.subjectivity

df['polarity'] = df['Message_punct'].apply(pol)
df['subjectivity'] = df['Message_punct'].apply(sub)
df.head()

Link  \
0  https://www.canadavisa.com//canada-immigration-discussion-board/threads/schengen-visa.656723/pos...   
1  https://www.canadavisa.com//canada-immigration-discussion-board/threads/visa-from-paris.636790/p...   
2  https://www.canadavisa.com//canada-immigration-discussion-board/threads/need-job-and-visa-in-can...   
3  https://www.canadavisa.com//canada-immigration-discussion-board/threads/need-job-and-visa-in-can...   
4  https://www.canadavisa.com//canada-immigration-discussion-board/threads/need-job-and-visa-in-can...   

        Date         Username  \
0 2019-09-30  Prashant-Canada   
1 2019-06-11      spanishnick   
2 2020-01-22     Junaid yasin   
3 2020-01-22     Junaid yasin   
4 2020-01-23     canuck_in_uk   

                                                                                               Message  \
0  I’m filing  Switzerland Visa for my family as a tourist .<br/><br/><br/><ul><li data-xf-list-typ...   
1  Hi,<br/><br/>My aunt and uncle applied for a tourist visa at the Canadian embassy in Paris more ...   
2                                                                                  I need job in canda   
3  <blockquote class="bbCodeBlock bbCodeBlock--expandable bbCodeBlock--quote"><div class="bbCodeBlo...   
4  can possible without IELTS test LMIA i am from Pakistan but working Saudi Arabia on Riyadh Metro...   

  Country Like                        Topic                      Thread  \
0     NaN    0  Visa Offices in Europe\nNew               Schengen VISA   
1     NaN    0  Visa Offices in Europe\nNew             Visa from Paris   
2     NaN    0  Visa Offices in Europe\nNew  Need job and visa in canda   
3     NaN    0  Visa Offices in Europe\nNew  Need job and visa in canda   
4     NaN    1  Visa Offices in Europe\nNew  Need job and visa in canda   

  month_year   forum    year  \
0    2019-09  Cforum  2019.0   
1    2019-06  Cforum  2019.0   
2    2020-01  Cforum  2020.0   
3    2020-01  Cforum  2020.0   
4    2020-01  Cforum  2020.0   

                                                                                     Message_lowercase  \
0  i’m filing switzerland visa for my family as a tourist .<br/><br/><br/><ul><li data-xf-list-type...   
1  hi,<br/><br/>my aunt and uncle applied for a tourist visa at the canadian embassy in paris more ...   
2                                                                                  i need job in canda   
3  <blockquote class="bbcodeblock bbcodeblock--expandable bbcodeblock--quote"><div class="bbcodeblo...   
4  can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...   

                                                                                     Message_no_quotes  \
0  i’m filing switzerland visa for my family as a tourist .<br/><br/><br/><ul><li data-xf-list-type...   
1  hi,<br/><br/>my aunt and uncle applied for a tourist visa at the canadian embassy in paris more ...   
2                                                                                  i need job in canda   
3  <a class="link link--internal" href="https://www.canadavisa.com/canada-immigration-discussion-bo...   
4  can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...   

                                                                                         Message_punct  \
0  i m filing switzerland visa for my family as a tourist brbrbrulli dataxflisttypeuli have an appo...   
1  hibrbrmy aunt and uncle applied for a tourist visa at the canadian embassy in paris more than fi...   
2                                                                                  i need job in canda   
3  blockquote classbbcodeblock bbcodeblockexpandable bbcodeblockquotediv classbbcodeblocktitlea cla...   
4  can possible without ielts test lmia i am from pakistan but working saudi arabia on riyadh metro...   

   polarity  subjectivity  
0  0.033333      0.282143  
1  0.226667   

In [ ]:
#compute sentiment scores (polarity) per message from 2016

sentiment_scores_tb = [round(TextBlob(messages).sentiment.polarity, 3) for messages in df['Message_punct']]
sentiment_category_tb = ['positive' if score > 0 
                             else 'negative' if score < 0 
                                 else 'neutral' 
                                     for score in sentiment_scores_tb]

#add column with sentiment category (positive/negative/neutral) - based on the Textblob library. 

df['sentiment'] = sentiment_category_tb 


In [ ]:
df.head()

Total no. of postive, negative & neutral messages and their comparative percentages across both forums

In [ ]:

PositiveSent = df.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(df)*100
print(pctPos)
NegativeSent = df.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(df)*100
print(pctNeg)
NeutralSent = df.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(df)*100
print(pctNeut)

import matplotlib.pyplot as plt

labels = 'positive', 'negative', 'neutral'
sizes = [pctPos, pctNeg, pctNeut]

fig1, ax1 = plt.subplots()
ax1.pie(sizes, labels=labels, autopct='%1.1f%%',
        shadow=True, startangle=90)
#ax1.axis('equal')  # Equal aspect ratio ensures that pie is drawn as a circle.

## Sentiment by forum
Couldnt figure out a more elegant way to do this. 

In [ ]:
PositiveSent = dfA.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(dfA)*100
print(pctPos)
NegativeSent = dfA.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(dfA)*100
print(pctNeg)
NeutralSent = dfA.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(dfA)*100
print(pctNeut)

import matplotlib.pyplot as plt

labels = 'positive', 'negative', 'neutral'
sizes = [pctPos, pctNeg, pctNeut]

fig1, ax1 = plt.subplots()
ax1.pie(sizes, labels=labels, autopct='%1.1f%%',
        shadow=True, startangle=90)

In [ ]:
PositiveSent = dfC.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(dfC)*100
print(pctPos)
NegativeSent = dfC.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(dfC)*100
print(pctNeg)
NeutralSent = dfC.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(dfC)*100
print(pctNeut)

import matplotlib.pyplot as plt

labels = 'positive', 'negative', 'neutral'
sizes = [pctPos, pctNeg, pctNeut]

fig1, ax1 = plt.subplots()
ax1.pie(sizes, labels=labels, autopct='%1.1f%%',
        shadow=True, startangle=90)

## Sentiment by Year

For each year compute the sentiment for Aforum and Cforum. Again - I think there would be a more elegant way to do this. I have just sorted by forum and then sliced into year groups.

Sliced this dataset but couldnt figure out the best way to use it considering all column values are uneven - i.e years = 5, forums = 2, and sentiment = 3.

In [ ]:
df_a = df[['year', 'forum', 'sentiment']]

df_a.head()



In [ ]:
y2016 = df[df.year == 2016]
PositiveSent = y2016.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2016)*100
print(pctPos)
NegativeSent = y2016.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2016)*100
print(pctNeg)
NeutralSent = y2016.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2016)*100
print(pctNeut)

y2017 = df[df.year == 2017]
PositiveSent = y2017.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2017)*100
print(pctPos)
NegativeSent = y2017.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2017)*100
print(pctNeg)
NeutralSent = y2017.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2017)*100
print(pctNeut)

y2018 = df[df.year == 2018]
PositiveSent = y2018.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2018)*100
print(pctPos)
NegativeSent = y2018.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2018)*100
print(pctNeg)
NeutralSent = y2018.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2018)*100
print(pctNeut)

y2019 = df[df.year == 2019]
PositiveSent = y2019.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2019)*100
print(pctPos)
NegativeSent = y2019.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2019)*100
print(pctNeg)
NeutralSent = y2019.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2019)*100
print(pctNeut)

y2020 = df[df.year == 2020]
PositiveSent = y2020.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2020)*100
print(pctPos)
NegativeSent = y2020.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2020)*100
print(pctNeg)
NeutralSent = y2020.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2020)*100
print(pctNeut)

In [ ]:
#Aforum
PositiveSent = y2016A.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2016A)*100
print(pctPos)
NegativeSent = y2016A.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2016A)*100
print(pctNeg)
NeutralSent = y2016A.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2016A)*100
print(pctNeut)


PositiveSent = y2017A.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2017A)*100
print(pctPos)
NegativeSent = y2017A.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2017A)*100
print(pctNeg)
NeutralSent = y2017A.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2017A)*100
print(pctNeut)


PositiveSent = y2018A.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2018A)*100
print(pctPos)
NegativeSent = y2018A.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2018A)*100
print(pctNeg)
NeutralSent = y2018A.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2018A)*100
print(pctNeut)


PositiveSent = y2019A.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2019A)*100
print(pctPos)
NegativeSent = y2019A.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2019A)*100
print(pctNeg)
NeutralSent = y2019A.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2019A)*100
print(pctNeut)


PositiveSent = y2020A.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2020A)*100
print(pctPos)
NegativeSent = y2020A.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2020A)*100
print(pctNeg)
NeutralSent = y2020A.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2020A)*100
print(pctNeut)

Nothing interesting happening with sentiments over year in Aforum. Pretty consistent over time. 

In [ ]:
#Cforum

PositiveSent = y2016C.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2016C)*100
print(pctPos)
NegativeSent = y2016C.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2016C)*100
print(pctNeg)
NeutralSent = y2016C.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2016C)*100
print(pctNeut)


PositiveSent = y2017C.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2017C)*100
print(pctPos)
NegativeSent = y2017C.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2017C)*100
print(pctNeg)
NeutralSent = y2017C.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2017C)*100
print(pctNeut)


PositiveSent = y2018C.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2018C)*100
print(pctPos)
NegativeSent = y2018C.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2018C)*100
print(pctNeg)
NeutralSent = y2018C.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2018C)*100
print(pctNeut)


PositiveSent = y2019C.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2019C)*100
print(pctPos)
NegativeSent = y2019C.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2019C)*100
print(pctNeg)
NeutralSent = y2019C.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2019C)*100
print(pctNeut)


PositiveSent = y2020C.apply(lambda x: True if 'positive' in list(x) else False, axis=1)
numOfPositiveRows = len(PositiveSent[PositiveSent == True].index)
pctPos = numOfPositiveRows/len(y2020C)*100
print(pctPos)
NegativeSent = y2020C.apply(lambda x: True if 'negative' in list(x) else False, axis=1)
numOfNegativeRows = len(NegativeSent[NegativeSent == True].index)
pctNeg = numOfNegativeRows/len(y2020C)*100
print(pctNeg)
NeutralSent = y2020C.apply(lambda x: True if 'neutral' in list(x) else False, axis=1)
numOfNeutralRows = len(NeutralSent[NeutralSent == True].index)
pctNeut = numOfNeutralRows/len(y2020C)*100
print(pctNeut)


## Findings
Nothing interesting happening over the 5 years for forums in general or individually. The sentiment is pretty consistent throughout the years. The comparision between the two however, is interesting - i.e Aforum overall, is more positive than Cforum. (Texblob just uses averages, maybe a different sentiment analysis library would be more useful). 

In [ ]:
#top 100 words by forum - Canada

#can = df[df.forum == 'Cforum']
#can.head()

#tf_idf_vector = TfidfVectorizer(ngram_range=(1,3), max_features=100)
#tf_idf_vector.fit_transform(can['Message_lemmatized'])
#tfidf_top_100 = tf_idf_vector.get_feature_names()
#print(tfidf_top_100)

In [ ]:
#top 100 words by forum - Australia
#aus = df[df.forum == 'Aforum']
#aus.head()

#tf_idf_vector = TfidfVectorizer(ngram_range=(1,4), max_features=100)
#tf_idf_vector.fit_transform(aus['Message_lemmatized'])
#tfidf_top_100 = tf_idf_vector.get_feature_names()
#print(tfidf_top_100)

In [ ]:
#top 100 words for 2016 in Australia
#aus2016 = aus[aus.year == 2016]
#aus2016.head()

#tf_idf_vector = TfidfVectorizer(ngram_range=(1,3), max_features=100)
#tf_idf_vector.fit_transform(aus2016['Message_lemmatized'])
#fidf_top_100 = tf_idf_vector.get_feature_names()
#print(tfidf_top_100)

In [ ]:
#top 100 words by sentiment and Australia
#auspositive = aus[aus.sentiment == 'positive']

#tf_idf_vector = TfidfVectorizer(ngram_range=(1,1), max_features=100)
#tf_idf_vector.fit_transform(auspositive['Message_lemmatized'])
#tfidf_top_100 = tf_idf_vector.get_feature_names()
#rint(tfidf_top_100)


In [ ]:
#top 100 words by sentiment and Australia
#ausnegative = aus[aus.sentiment == 'negative']

#tf_idf_vector = TfidfVectorizer(ngram_range=(1,1), max_features=100)
#tf_idf_vector.fit_transform(ausnegative['Message_lemmatized'])
#tfidf_top_100 = tf_idf_vector.get_feature_names()
#print(tfidf_top_100)

#words like 'bridging', 'sorry', 'student', 'wrong' not appearing in positive sentiment dataset. 

In [ ]:
#top 100 n word phrases
#tf_idf_vector = TfidfVectorizer(ngram_range=(10,10), max_features=100)
#tf_idf_vector.fit_transform(ausnegative['Message'])
#tfidf_top_100 = tf_idf_vector.get_feature_names()
#print(tfidf_top_100)

In [ ]:
#tf_idf_vector = TfidfVectorizer(ngram_range=(4,4), max_features=100)
#tf_idf_vector.fit_transform(auspositive['Message'])
#tfidf_top_100 = tf_idf_vector.get_feature_names()
#print(tfidf_top_100)

In [ ]:
#canpositive = can[can.sentiment == 'positive']
#tf_idf_vector = TfidfVectorizer(ngram_range=(1,4), max_features=100)
#tf_idf_vector.fit_transform(canpositive['Message_lemmatized'])
#tfidf_top_100 = tf_idf_vector.get_feature_names()
#print(tfidf_top_100)

In [ ]:
#cannegative = can[can.sentiment == 'negative']
#tf_idf_vector = TfidfVectorizer(ngram_range=(3,4), max_features=100)
#tf_idf_vector.fit_transform(cannegative['Message_lemmatized'])
#tfidf_top_100 = tf_idf_vector.get_feature_names()
#rint(tfidf_top_100)



In [ ]:
#len(cannegative)

In [ ]:
#len(canpositive)

In [ ]:
#len(ausnegative)

In [ ]:
#len(auspositive)